rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+
+### Emphasis on flake over shell script wrappers
+
+[nixos-rebuild](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh) is nothing but a complex shell script, that has to be present in the system beforehand, which basically calls `nix-build` under the hood, building a derivation, and then running the built derivation. All the options can be instead defined in nix language, in the flake definition itself, and the output derivation can be built using the nix command directly.
+
+[nixos-generators](https://github.com/nix-community/nixos-generators#using-in-a-flake) has quite recently started adding flake support through which there's no need to install or do anything of it, a simple `nix build` will build the iso/vm images. Making it transparent as well as reducing the extra script installation setup.
+
+### CLI improvements
+
+This is majorly for improving the binary cache user interaction, remote-builders setup, and cache serving.
+
+This is not the exact design required, but something like this will suffice:
+
+```bash
+nix-store -r /nix/store/{some.drv}        ## OR: nix build /nix/store/{some.drv}
+# pulls/builds whatever is in $(nix show-derivation /nix/store/{some.drv} | jq '.[] | .inputSrcs, .inputDrvs')
+# builds and produces output
+
+nix-store --export /nix/store/{somepath} > file
+
+## This is the main point, pull paths from binary cache, if we exactly know it
+nix store pull $(nix-store --query --requisites /nix/store/{some.drv})
+nix-store --import < file
+```
+
+The purpose of remote builder shouldn't be exactly to serve the built file immediately, maybe to store it in some static-storage, and shutdown, the cache serving should be made accessible using stateless http server, which can happen if we can serve a folder through http.
+
+Articles confuse the binary cache with closure copying, such as [this one](https://nix-tutorial.gitlabpages.inria.fr/nix-tutorial/deploying.html), one of top results in Google with "setting up a binary cache nix" query.
+
+Hence there shouldn't be need of third party package in order to serve something, unnecessary complexity & abstractions should be reduced to fit a simple custom http file server to act as binary cache.
+
+```bash
+# Something like:
+nix store setup > /nix/store/nix-cache-info
+
+python3 -m http.server -d /nix/store
+```
+
+An alternative way could be use of a flake to build isolated directory with the filtered store path references only, or even an option to have upstream filtered paths (that are not in https://cache.nixos.org). Something like:
+
+```nix
+{
+  # ...
+  outputs = args: {
+    packages.x86_64-linux.my-binary-cache = some-binary-cache-setup-function {
+      packages = [ custom-package-1 custom-package-2 ];
+      # ...
+    };
+  };
+}
+
+# nix build '.#my-binary-cache'
+# python3 -m http.server -d result
```

Comment:
`nix build .# --store file:///$(pwd)/result`? Or if you don't want the build closure in there, `nix build .#; nix copy ./result --to file:///$(pwd)/cache`? (haven't tested the exact commands though).

Reply:
From the description https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-copy.html:

`nix copy` copies store path closures between two Nix stores.

I don't think there's a ingenious way to do this in the new cli at the moment, the closest is the `nix-store --export`...